src/Stratis.Bitcoin.Features.Miner/PosMinting.cs

```diff
@@ -402,14 +408,37 @@ public IAsyncLoop Stake(WalletSecret walletSecret)
 
                 this.logger.LogTrace("(-)");
             },
-            this.nodeLifetime.ApplicationStopping,
+            this.stakeCancellationTokenSource.Token,
             repeatEvery: TimeSpan.FromMilliseconds(this.minerSleep),
             startAfter: TimeSpans.TenSeconds);
 
             this.logger.LogTrace("(-)");
             return this.stakingLoop;
         }
 
+        /// <summary>
+        /// Stop the main POS staking loop.
+        /// </summary>
+        public void StopStake()
+        {
+            this.logger.LogTrace("()");
+
+            if (this.stakingLoop == null)
+            {
+                this.logger.LogTrace("(-)[NOT_MINING]");
+                return;
+            }
+
+            this.stakeCancellationTokenSource.Cancel();
+            this.stakingLoop.Dispose();
+            this.stakingLoop = null;
+            this.stakeCancellationTokenSource = null;
```

Comment:
how about disposing first?

Reply:
Disposing blocks

Reply:
what?

Reply:
Here: https://docs.microsoft.com/en-us/dotnet/standard/threading/cancellation-in-managed-threads it says:

> The CancellationTokenSource class implements the IDisposable interface. You should be sure to call the CancellationTokenSource.Dispose method when you have finished using the cancellation token source to free any unmanaged resources it holds.

Reply:
No but I mean the dispose implementation of the async loop blocks until the loop ends.
So we have to cancel first then wait for the loop to finish calling the Dispose()
Then we can dispose the cancellation.